### ISSUE TITLE
API consistency

### ISSUE THREAD

**Issue by tumbl3w33d** [kanidm/kanidm#1954]
I noticed an inconsistency in the API when it comes to the data type of `gidnumber`. When retrieving a group via `GET` to `/v1/group`, the `gidnumber` is returned as (list of) **string**, however, when you `POST` to the unix config of that group via `/v1/group/:id/_unix`, it asks you for an (optional) **integer**. I would have expected to `POST` an empty list instead of `null` or use `DELETE`.

It's just a minor thing, hardly worth opening the issue for and it reminds me of the not so perfect status codes the API returns (see #1904). Would you prefer to ignore such things, or have an issue per finding or rather collect everything in a long-lasting "API improvement" issue?

Edit: I actually didn't yet figure out how to delete the `gidnumber` from a group. Neither `DELETE`, nor `POST`ing `null` works, even though the latter returns `200`, because well, the integer is optional, it just doesn't do anything. 😅

**Comment by yaleman** [comment]
Any improvement to dev consistency and ergonomics is *always* worth doing - it's just whether we've hit that use case yet 😄 

1. The `GET` verb probably should return a single integer, but the default response of the database is a list; I'd say that could be nicer, and we should change that response (since there should only ever be one `gidnumber` for one `group`.
2. I'm not sure what the expectation is for `POST`ing a null to that `/v1/group/:id/_unix` endpoint - it feels like we should be requiring that value or throwing an error TBH.
3. Removing the `gidnumber` could be a `DELETE` for sure - I guess we just haven't used that yet. 😄 - have you tried the `/v1/group/:id/_attr/:attr` which has a `DELETE` verb?

**Comment by tumbl3w33d** [comment]
>  the default response of the database is a list; I'd say that could be nicer, and we should change that response (since there should only ever be one gidnumber for one group

I thought that's the general pattern. Pretty much everything I've seen is returned as list by the API, no matter how likely it is that more than one is returned. If you would want to change that, I guess a lot of code/interface would change. I'm neutral on that and got used to handle it in my client.

> I'm not sure what the expectation is for POSTing a null to that /v1/group/:id/_unix endpoint - it feels like we should be requiring that value or throwing an error TBH.

Agree. I'm no fan of associating `null` with any meaning. It was just me trying to get rid of the gidnumber (without success).

> Removing the gidnumber could be a DELETE for sure - I guess we just haven't used that yet. smile - have you tried the /v1/group/:id/_attr/:attr which has a DELETE verb?

Good idea, I tried that. The behavior is interesting, in some way. So one aspect is, that I find it weird to use the `unix` path for creating/modifying and the `attr` path for deleting, because, why not setting it via `attr` path in the first place? Apart from that, the `gidnumber` has a special behavior attached to it:

When you delete the `gidnumber` that you set earlier, it really gets removed, however, you don't end up without any but instead you receive the "randomly" generated one. You can then override it again by setting an own. Then my question would be - how do I remove the entire posix class from the group? I guess for this purpose the `DELETE` to the `unix` path would be appropriate. For groups that would make sense, because AFAIK they only have 1 unix attribute. `person` and `service-account` have more  (2 that I know) so a DELETE would need a payload ([meh](https://stackoverflow.com/questions/299628/is-an-entity-body-allowed-for-an-http-delete-request)) to define what shall be deleted or we are back to `POST`/`PUT` which overwrites the entire state, i.e., the payload contains the full list of attributes to set and everything else gets discarded.

**Comment by Firstyear** [comment]
@tumbl3w33d Some of this inconsistency is because of evolution - when I started with the api, I didn't realise all the edge cases and how it should look or feel to a user. So there will be plenty of rough edges. But right now the api isn't *stable* so we can make as many changes as we like.

The reason in this case for the error is that there is a "generic" attribute post/get api (`/v1/<class>/_attr/<attr>`) that treats everything as strings (internally it's a ProtoValue from a ProtoEntry, and that's converted in https://github.com/kanidm/kanidm/blob/master/server/lib/src/server/mod.rs#L501 from a string to an internally strong typed value. 

However as you notice there is also a *strongly typed* variant of this api certain class-magic end points. This is things like `/v1/<class>/_unix` and because of that these interfaces actually know what types things should be and that's why it takes an int here. 

I think for now we can't fix this because the generic version always has to be ... well generic. But I think we *should* move toward #211 instead which does solve this and presents a better api for us. And in that process we should retire some of our generic apis and leave magic handlers under the scim interface. How does that sound?

**Comment by tumbl3w33d** [comment]
Now that you told me about the idea behind it, with generic vs. strongly typed API, it's actually good enough for me.

The root cause seems to be

-  me not knowing what I'm doing, stumbling through the rust code to read what the API offers plus 
-  me not knowing about this distinction of APIs and which are the preferred ones to use for what I'm trying to achieve

This in combination with the lack of helpful error codes (or lack of error response at all) makes you end up confused.

P.S.: I noticed that some of the API code lines I was looking at were 3-4 years old. I sometimes can't stand my code after 2 weeks, so big respect for how good it still looks and works in kanidm.

**Comment by Firstyear** [comment]
So I think part of that is #1984 where we need better communication of errors. But yeah, there is a lot to improve here. I think generic-but-typed is what we need, so scim will be a big win here for you. Would you be okay with us considering this issue to be "pls impl scim" because it resolves the consistency and type issues?

**Comment by tumbl3w33d** [comment]
I will not nit-pick about the generic API anymore, but I guess for the typed ones I can still complain, right? :)

I started to use `/v1/oauth2` and everything below and noticed a weird type expectation when `POST`ing the following payload to `/v1/oauth2/_basic`.
```json
{
    "attrs": {
        "oauth2_rs_name": [
            "my-app"
        ],
        "displayname": [
            "My App"
        ],
        "oauth2_rs_origin": [
            "http://localhost:5000/"
        ],
        "oauth2_allow_insecure_client_disable_pkce": [false],
        "oauth2_prefer_short_username": [true]
    }
}
```
The result is a `422` from the server and the complaint:
```
Failed to deserialize the JSON body into the target type: attrs.oauth2_allow_insecure_client_disable_pkce[0]: invalid type: boolean `false`, expected a string
```
While I got used to wrapping everything in lists, it's really strange to turn my boolean into a list of strings before `POST`ing.

Then I tried to add the `oauth2_rs_basic_secret` to the payload and got another strange result:
```
500 - b'{"invalidattribute":"Radius secrets can not be supplied through modification - please use the IDM api"}'
```
I don't see a connection to anything Radius here, so I guess this is a copy-paste error message. So I assume it's not wanted to set the secret on creation. I don't see why, but okay, I would go ahead and expect `/v1/oauth2/:rs_name/_basic_secret` to  eat a `POST` but it only supports `GET`. I will try with `PATCH` against `/v1/oauth2/:rs_name` yet but got the feeling, it's currently not possible to manually set the client secret.

**Comment by yaleman** [comment]
We don't allow setting OAuth secrets. You can reset them or get them. See #1747

**Comment by tumbl3w33d** [comment]
Alright, I remember we talked about something like that, but I thought it was only about personal credentials. No big deal - I already adjusted my logic to just gather the secrets from the API and store them in a git-cryped file in my ansible repo.

**Comment by tumbl3w33d** [comment]
Disclaimer: Please let me know when I'm abusing this ticket too much as sink for my findings.

The OAuth2 API returns the scope map in a rather unfortunate format, which is probably fine for your internal use because you share the serializing/deserializing logic among tools. From the outside to parse a response like this:

```json
{ "oauth2_rs_scope_map": ["administrators@idm.example.com: {\"admin\", \"displayname\", \"email\", \"preferred_username\", \"profile\"}', 'internal@idm.example.com: {\"oidc\"}"]}
```
it takes quite some action:
``` python
    api_scope_map = {}
    for item in kanidm_oauth2['attrs']['oauth2_rs_scope_map']:
        group, scopes = get_group_and_scopes_from_api_response(item)
        DISPLAY.v(f"scopes of group {group}: {scopes}")
        api_scope_map[group] = scopes

[…]

def get_group_and_scopes_from_api_response(scope_map_resp):
    # Split the string based on ' : '
    spn, scopes_str = scope_map_resp.split(': ')

    # Extract group name from the key part (which is the group's spn)
    group = spn.split('@')[0]

    # Remove surrounding '{}' from the scopes string and split by ','
    scope_list = scopes_str.strip('{}').split(', ')

    # Remove the quotes from the scopes and create a set
    scopes = {scope.strip('"') for scope in scope_list}
    return group, scopes
```
which is certainly doable, but it adds up.

What it should look like:
```json
{ "oauth2_rs_scope_map": { "administrators@idm.example.com" : ["admin", "displayname", "email", "preferred_username", "profile"], "internal@idm.example.com": ["oidc"]}}
```

**Comment by yaleman** [comment]
> Disclaimer: Please let me know when I'm abusing this ticket too much as sink for my findings.

We really need to start breaking it up into smaller related chunks 😄

**Comment by Firstyear** [comment]
Yeah, I think we need to focus on the core scim bits first then look at migrating over admin features to scim-like setups.
